### PR TITLE
SC2: Add Roach Mercenaries

### PR DESCRIPTION
**Pull request by Snarkie**
## What is this fixing or adding?

Adds a Roach Mercenary item. Working name: Thornshells.
Stronger versions of regular Roaches, that can attack air.

## How was this tested?

Generated a local campaign, added item with console commands

~~Note, that there is currently some issue related to Pygalisk, you will unlock both this merc and Pygalisks at the same time. Pygalisks are already being changed at this time.~~
This has been resolved

[Client PR](https://github.com/Ziktofel/Archipelago/pull/363)

### CODE REVIEW

**Comment by Ziktofel** [comment]
Snímek obrazovky pořízený 2024-12-05 11-51-52
Snímek obrazovky pořízený 2024-12-05 11-51-44
Check Malignant Creep bonus

**Review by Ziktofel — CHANGES_REQUESTED** [review]
I'll check training after the Pygalisk related change gets merged

**Comment by Snarkie** [comment]
> Check Malignant Creep bonus
fixed

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UpgradeData.xml [hunk 1]
```diff
@@ -2400,14 +2421,14 @@
...
         <EffectArray Reference="Weapon,AP_RoachCorpserMelee,Level" Value="1"/>
-        <EffectArray Reference="Effect,AP_RoachCorpserAcidSalivaU,Amount" Value="2"/>
+        <EffectArray Reference="Effect,AP_RoachCorpserAcidSalivaU,Amount" Value="3"/>
         <EffectArray Reference="Weapon,AP_RoachCorpserAcidSaliva,Level" Value="1"/>
         <EffectArray Reference="Weapon,AP_RoachVileCorpserAcidSaliva,Level" Value="1"/>
-        <EffectArray Reference="Effect,AP_RoachVileUMelee,Amount" Value="2"/>
+        <EffectArray Reference="Effect,AP_RoachVileUMelee,Amount" Value="3"/>
         <EffectArray Reference="Weapon,AP_RoachVileMelee,Level" Value="1"/>
-        <EffectArray Reference="Effect,AP_RoachVileAcidSalivaU,Amount" Value="2"/>
+        <EffectArray Reference="Effect,AP_RoachVileAcidSalivaU,Amount" Value="3"/>
```
[the remark this answers is not in the document]
Comment (Snarkie) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UpgradeData.xml]: Correct, that was unintended. Fixed.

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/ActorData.xml [hunk 2]
```diff
@@ -42259,4 +42259,212 @@
...
+    <CActorMissile id="AP_MercRoachAttackMissile" parent="AP_RoachAttackMissile" unitName="AP_MercRoachAcidSalivaWeapon">
+        <Model value="AP_MercRoachAcidSalivaWeapon"/>
+        <On index="0" Terms="UnitBirth.AP_MercRoachAcidSalivaWeapon" Send="Create"/>
+        <On Terms="ActorCreation" Send="SetTintColor {255,0,255 2.000000}"/>
+        <Scale value="1.500000"/>
+    </CActorMissile>
+    <CActorAction id="AP_MercRoachCorpserAttack" parent="AP_RoachCorpserAttack" effectImpact="AP_MercRoachCorpserAcidSalivaU" effectLaunch="AP_MercRoachCorpserAcidSalivaLM">
+        <On index="1" Terms="Effect.AP_MercRoachCorpserAcidSalivaU.Start; FromEffectTreeDescendant"/>
+        <On index="2" Terms="Effect.AP_MercRoachCorpserAcidSalivaU.Stop; FromEffectTreeDescendant"/>
+        <On index="6" Terms="Effect.AP_MercRoachCorpserAcidSalivaLM.Start; At Caster" Send="Create"/>
```
[the remark this answers is not in the document]
Comment (Snarkie) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/ActorData.xml]: All event indexes replaced with Remove + On

**Comment by Ziktofel** [comment]
Fixed the attack missile spawning point

**Comment by Snarkie** [comment]
> Fixed the attack missile spawning point

Funny, my initial setup had something very similar (I went with .7 offset)

Ultimately decided against it because regular roach missile also awkwardy starts more towards the center of the unit and I thought I'd keep it consistent.

**Review by Ziktofel — CHANGES_REQUESTED** [review]

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml [hunk 3]
```diff
@@ -4202,4 +4202,142 @@
...
+    <CWeaponLegacy id="AP_MercRoachVileAcidSalivaAir">
+        <EditorCategories value="Race:Zerg"/>
+        <Tip value="Weapon/Tip/AP_MercRoachVileAcidSaliva"/>
+        <Options index="Disabled" value="1"/>
+        <Options index="Hidden" value="1"/>
+        <Icon value="Assets\Textures\btn-upgrade-zerg-missileattacks-level0.dds"/>
+        <DisplayEffect value="AP_MercRoachVileAcidSalivaU"/>
+        <TargetFilters value="Air,Visible;Missile,Stasis,Dead,Hidden,Invulnerable"/>
+        <Range value="4"/>
+        <MinimumRange value="0.6"/>
```
[the remark this answers is not in the document]
Comment (Snarkie) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml]: fixed

**Comment by Snarkie** [comment]
Tested again in local campaign. 

- Training works
- air attack close range works
- actors seem to work
- all items work (including strain combinations)
- malignant creep works
- mercenary cooldown upgrades work

**Pull request merged**